### PR TITLE
Kafka - Event Bus

### PR DESCRIPTION
**Pull request by BohdanRysakov**

### CODE REVIEW

**Comment by BohdanRysakov** [comment]
@ajax-shulepa-a 
@ajax-sushkevych-s 
@ajax-gnatiuk-s

File: .github/workflows/github-actions.yaml [hunk 1]
```diff
@@ -41,5 +32,5 @@ jobs:
       - name: Grant execute permission for gradlew
         run: chmod +x gradlew
 
-      - name: Compile and run checks
-        run: ./gradlew build
+      - name: Compile
+        run: ./gradlew assemble
```
Comment (ajax-shulepa-a) [inline .github/workflows/github-actions.yaml:36]: why did you switch to assemble task>

File: api/src/main/proto/input/reqreply/message/create.proto [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+message MessageCreateResponse {
+
+  oneof response {
+    Success success = 1;
+    Failure failure = 2;
+  }
+
+  message Success {
+    string result = 1;
```
Comment (ajax-shulepa-a) [inline api/src/main/proto/input/reqreply/message/create.proto:23]: what is purpose of this string?

File: api/src/main/proto/input/reqreply/message/create.proto [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  message Failure {
+    optional string message = 1;
+    commonmodels.error.proto.Error error = 2;
```
Comment (ajax-shulepa-a) [inline api/src/main/proto/input/reqreply/message/create.proto:28]: the goal is to have here a list of business errors, general 'error' doesn't say anything about what happened

File: api/src/main/proto/input/reqreply/message/eventSubscription.proto [hunk 4]
```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
+
+package rys.ajaxpetproject.request.message.subscription.proto;
+
+import "commonmodels/message/messageDto.proto";
+import "commonmodels/error/error.proto";
+
+message CreateSubscriptionRequest {
+  repeated string chat = 1;
```
Comment (ajax-shulepa-a) [inline api/src/main/proto/input/reqreply/message/eventSubscription.proto:9]: chatId?

File: api/src/main/proto/input/reqreply/message/messageGrpcService.proto [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import "input/reqreply/message/create.proto";
+import "input/reqreply/message/eventSubscription.proto";
+
+package rys.ajaxpetproject.service.message;
+
+service MessageService {
+  rpc create(request.message.create.proto.MessageCreateRequest)
+      returns (request.message.create.proto.MessageCreateResponse);
+
+  rpc subscribe(stream request.message.subscription.proto.CreateSubscriptionRequest)
```
Comment (ajax-shulepa-a) [inline api/src/main/proto/input/reqreply/message/messageGrpcService.proto:12]: stream? and StreamMessagesRequest/Response

File: gRPC/src/main/kotlin/rys/ajaxpetproject/grpc/MessageGrpc.kt [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+import rys.ajaxpetproject.utils.toModel
+import rys.ajaxpetproject.internalapi.MessageEvent
+import rys.ajaxpetproject.model.MongoMessage
+import rys.ajaxpetproject.request.message.subscription.proto.EventSubscription.CreateSubscriptionRequest
+import rys.ajaxpetproject.request.message.subscription.proto.EventSubscription.CreateSubscriptionResponse
+import rys.ajaxpetproject.utils.toProto
+
+@GrpcService
+@Suppress("TooGenericExceptionCaught")
+class MessageGrpc(
```
Comment (ajax-shulepa-a) [inline gRPC/src/main/kotlin/rys/ajaxpetproject/grpc/MessageGrpc.kt:25]: MessageGrpcService

your class is not 'remote procedure call', right?

File: gRPC/src/main/kotlin/rys/ajaxpetproject/grpc/MessageGrpc.kt [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    natsConnection: Connection,
+    private val chatService: ChatService,
+    private val messageService: MessageService,
+) : ReactorMessageServiceGrpc.MessageServiceImplBase() {
+
+    private val messageParser = MessageDto.parser()
+    private val dispatcher = natsConnection.createDispatcher()
+
+    override fun subscribe(requests: Flux<CreateSubscriptionRequest>): Flux<CreateSubscriptionResponse> {
```
Comment (ajax-shulepa-a) [inline gRPC/src/main/kotlin/rys/ajaxpetproject/grpc/MessageGrpc.kt:34]: let's discuss in TG why you need a bidirectional stream here

File: gRPC/src/main/kotlin/rys/ajaxpetproject/grpc/MessageGrpc.kt [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+@GrpcService
+@Suppress("TooGenericExceptionCaught")
+class MessageGrpc(
+    natsConnection: Connection,
+    private val chatService: ChatService,
+    private val messageService: MessageService,
+) : ReactorMessageServiceGrpc.MessageServiceImplBase() {
+
+    private val messageParser = MessageDto.parser()
+    private val dispatcher = natsConnection.createDispatcher()
```
Comment (ajax-shulepa-a) [inline gRPC/src/main/kotlin/rys/ajaxpetproject/grpc/MessageGrpc.kt:32]: you're mixing grpc and nats here, consider wrap interaction with nats in some class

**Review by ajax-shulepa-a — CHANGES_REQUESTED** [review]

File: .github/workflows/github-actions.yaml (hunk 1 above)
Comment (BohdanRysakov) [reply .github/workflows/github-actions.yaml:36]: Some issue with slf4. In tests gives this error :

SLF4J: Class path contains multiple SLF4J providers.
SLF4J: Found provider [org.slf4j.reload4j.Reload4jServiceProvider@38ba6ce3]
SLF4J: Found provider [ch.qos.logback.classic.spi.LogbackServiceProvider@d278d2b]
SLF4J: See https://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual provider is of type [org.slf4j.reload4j.Reload4jServiceProvider@38ba6ce3]

Working on this. Almost done

File: api/src/main/proto/input/reqreply/message/create.proto (hunk 2 above)
Comment (BohdanRysakov) [reply api/src/main/proto/input/reqreply/message/create.proto:23]: Give some success message back. 
Not necessary but logic was something like : "MessageService saved message and tells what he saved BUT I don't want Client to see other fields in Message"

File: api/src/main/proto/input/reqreply/message/create.proto (hunk 3 above)
Comment (BohdanRysakov) [reply api/src/main/proto/input/reqreply/message/create.proto:28]: Yes, but I will took time to write many errors and test them. So in order to not doing copy-paste actions I've left just error

File: api/src/main/proto/input/reqreply/message/createEvent.proto [hunk 9]
```diff
@@ -0,0 +1,10 @@
+syntax = "proto3";
+
+package rys.ajaxpetproject.request.message.create.proto;
+
+import "commonmodels/message/message.proto";
+
+message MessageCreateEvent {
```
[the remark this answers is not in the document]
Comment (BohdanRysakov) [reply api/src/main/proto/input/reqreply/message/createEvent.proto:7]: Thanks! 
Dont

File: core/src/main/kotlin/rys/ajaxpetproject/service/impl/ChatServiceImpl.kt [hunk 10]
```diff
@@ -1,32 +1,36 @@
...
+import rys.ajaxpetproject.utils.createEvent
 
 @Service
 @Suppress("TooManyFunctions")
 class ChatServiceImpl(
     private val chatRepository: ChatRepository,
     private val userService: UserService,
-    private val messageService: MessageService
+    private val messageService: MessageService,
+    private val kafka: MessageCreateEventProducer
```
[the remark this answers is not in the document]
Comment (BohdanRysakov) [reply core/src/main/kotlin/rys/ajaxpetproject/service/impl/ChatServiceImpl.kt:26]: Renamed for kafkaEventSender

File: core/src/main/kotlin/rys/ajaxpetproject/service/impl/ChatServiceImpl.kt [hunk 11]
```diff
@@ -60,13 +58,22 @@ class ChatServiceImpl(
...
+        return chatRepository.addMessage(messageId, chatId)
+            .flatMap {
+                messageService.getMessageById(messageId)
+                    .flatMap {
+                        kafka.sendCreateEvent(
+                            it.createEvent(chatId)
+                        )
+                    }.onErrorResume {
+                        Mono.error(it)
+                    }
```
[the remark this answers is not in the document]
Comment (BohdanRysakov) [reply core/src/main/kotlin/rys/ajaxpetproject/service/impl/ChatServiceImpl.kt:70]: removed

File: gRPC/src/main/kotlin/rys/ajaxpetproject/grpc/MessageGrpc.kt (hunk 6 above)
Comment (BohdanRysakov) [reply gRPC/src/main/kotlin/rys/ajaxpetproject/grpc/MessageGrpc.kt:25]: Yeah , not suitable name. Changed

File: api/src/main/proto/input/reqreply/message/eventSubscription.proto (hunk 4 above)
Comment (BohdanRysakov) [reply api/src/main/proto/input/reqreply/message/eventSubscription.proto]: Done

File: .github/workflows/github-actions.yaml (hunk 1 above)
Comment (BohdanRysakov) [reply .github/workflows/github-actions.yaml:28]: Done

File: api/src/main/proto/input/reqreply/message/create.proto (hunk 2 above)
Comment (ajax-shulepa-a) [reply api/src/main/proto/input/reqreply/message/create.proto:23]: Success is by it self marker that create was finished successfully, message there is redundant

File: core/src/main/kotlin/rys/ajaxpetproject/service/MessageEventService.kt [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+            }
+        }
+    }
+
+    fun loadInitialState(chatId: String): Flux<EventSubscription.CreateSubscriptionResponse> {
+        return chatService.getMessagesInChat(chatId)
+            .flatMap { message ->
+                val messageDto = message.toDto(chatId)
+                val response = buildSuccessResponse(messageDto)
+                Flux.just(response)
```
Comment (ajax-shulepa-a) [inline core/src/main/kotlin/rys/ajaxpetproject/service/MessageEventService.kt:50]: do you need Flux.just here?

**Review by ajax-shulepa-a — APPROVED** [review]

File: api/src/main/proto/input/reqreply/message/create.proto (hunk 2 above)
Comment (BohdanRysakov) [reply api/src/main/proto/input/reqreply/message/create.proto:23]: Fair enough, thanks

File: core/src/main/kotlin/rys/ajaxpetproject/service/MessageEventService.kt (hunk 12 above)
Comment (BohdanRysakov) [reply core/src/main/kotlin/rys/ajaxpetproject/service/MessageEventService.kt:50]: Nope, mono is enough, thanks

**Pull request merged**